### ISSUE TITLE
Layer-wise Spiking Activity

### ISSUE THREAD

**Issue by tehreemnaqvi** [guillaume-chevalier/Spiking-Neural-Network-SNN-with-PyTorch-where-Backpropagation-engenders-STDP#6]
I want to visualization the spiking activity for each layer and used your visualization functions.

But I got some errors.

Can you please tell me, where is the issue?

Below is my full code with the error:

class SCNN(nn.Module):
    def __init__(self, n_time_steps):
        super(SCNN, self).__init__()
        self.n_time_steps = n_time_steps
        in_planes, out_planes, stride, padding, kernel_size = cfg_cnn[0]
        self.conv1 = nn.Conv2d(in_planes, out_planes, kernel_size=kernel_size, stride=stride, padding=padding)
        in_planes, out_planes, stride, padding, kernel_size = cfg_cnn[1]
        self.conv2 = nn.Conv2d(in_planes, out_planes, kernel_size=kernel_size, stride=stride, padding=padding)

        self.fc1 = nn.Linear(cfg_kernel[-1] * cfg_kernel[-1] * cfg_cnn[-1][1], cfg_fc[0])
        self.fc2 = nn.Linear(cfg_fc[0], cfg_fc[1])

    def forward_through_time(self, input):
        outputs = []
        all_layer1_states = []
        all_layer1_outputs = []
        all_layer2_states = []
        all_layer2_outputs = []
        all_layer3_states = []
        all_layer3_outputs = []
        all_layer4_states = []
        all_layer4_outputs = []
        c1_mem = c1_spike = torch.zeros(batch_size, cfg_cnn[0][1], cfg_kernel[0], cfg_kernel[0], device=device)
        c2_mem = c2_spike = torch.zeros(batch_size, cfg_cnn[1][1], cfg_kernel[1], cfg_kernel[1], device=device)

        h1_mem = h1_spike = h1_sumspike = torch.zeros(batch_size, cfg_fc[0], device=device)
        h2_mem = h2_spike = h2_sumspike = torch.zeros(batch_size, cfg_fc[1], device=device)

        for step in range(self.n_time_steps):
            x = input > torch.rand(input.size(), device=device)  # prob. firing

            c1_mem, c1_spike = mem_update(self.conv1, x.float(), c1_mem, c1_spike)
            #print(c1_mem.shape)

            x = F.avg_pool2d(c1_spike, 2)

            c2_mem, c2_spike = mem_update(self.conv2, x, c2_mem, c2_spike)

            x = F.avg_pool2d(c2_spike, 2)
            x = x.view(batch_size, -1)

            h1_mem, h1_spike = mem_update(self.fc1, x, h1_mem, h1_spike)
            h1_sumspike += h1_spike
            h2_mem, h2_spike = mem_update(self.fc2, h1_spike, h2_mem, h2_spike)
            h2_sumspike += h2_spike

            all_layer1_states.append(c1_mem)
            all_layer1_outputs.append(c1_spike)
            all_layer2_states.append(c2_mem)
            all_layer2_outputs.append(c2_spike)
            all_layer3_states.append(h1_mem)
            all_layer3_outputs.append(h1_spike)
            all_layer4_states.append(h2_mem)
            all_layer4_outputs.append(h2_spike)
            outputs.append(h2_sumspike)

        outputs = h2_sumspike / self.n_time_steps
        return outputs, [[all_layer1_states, all_layer1_outputs], [all_layer2_outputs, all_layer2_outputs],
                         [all_layer3_states, all_layer3_outputs], [all_layer4_states, all_layer4_outputs]]

    def forward(self, x):
        out, _ = self.forward_through_time(x)
        return out

    def visualize_all_neurons(self, x):
        print(x.shape)
        assert x.shape[0] == 1 and len(x.shape) == 4, (
            "Pass only 1 example to SpikingNet.visualize(x) with outer dimension shape of 1.")
        _, layers_state = self.forward_through_time(x)

        for i, (all_layer_states, all_layer_outputs) in enumerate(layers_state):
            layer_state = torch.stack(all_layer_states).data.cpu().numpy().squeeze().transpose()
            layer_output = torch.stack(all_layer_outputs).data.cpu().numpy().squeeze().transpose()

            self.plot_layer(layer_state, title="Inner state values of neurons for layer {}".format(i))
            self.plot_layer(layer_output, title="Output spikes (activation) values of neurons for layer {}".format(i))

    def visualize_neuron(self, x, layer_idx, neuron_idx):
        assert x.shape[0] == 1 and len(x.shape) == 4, (
            "Pass only 1 example to SpikingNet.visualize(x) with outer dimension shape of 1.")
        _, layers_state = self.forward_through_time(x)

        all_layer_states, all_layer_outputs = layers_state[layer_idx]
        layer_state = torch.stack(all_layer_states).data.cpu().numpy().squeeze().transpose()
        layer_output = torch.stack(all_layer_outputs).data.cpu().numpy().squeeze().transpose()

        self.plot_neuron(layer_state[neuron_idx],
                         title="Inner state values neuron {} of layer {}".format(neuron_idx, layer_idx))
        self.plot_neuron(layer_output[neuron_idx],
                         title="Output spikes (activation) values of neuron {} of layer {}".format(neuron_idx,
                                                                                                   layer_idx))

    def plot_layer(self, layer_values, title):

        width = max(16, layer_values.shape[0] / 8)
        height = max(4, layer_values.shape[1] / 8)
        plt.figure(figsize=(width, height))
        plt.imshow(
            layer_values,
            interpolation="nearest",
            cmap=plt.cm.rainbow
        )
        plt.title(title)
        plt.colorbar()
        plt.xlabel("Time")
        plt.ylabel("Neurons of layer")
        plt.show()

    def plot_neuron(self, neuron_through_time, title):
        width = max(16, len(neuron_through_time) / 8)
        height = 4
        plt.figure(figsize=(width, height))
        plt.title(title)
        plt.plot(neuron_through_time)
        plt.xlabel("Time")
        plt.ylabel("Neuron's activation")
        plt.show()

print_freq = 500

train_accs = []
train_losses = []
val_losses = []
val_accs = []

best_acc = 0  # best test accuracy
start_epoch = 0  # start from epoch 0 or last checkpoint epoch
acc_record = list([])
loss_train_record = list([])
loss_test_record = list([])

snn = SCNN(n_time_steps=20)
snn.to(device)
criterion = nn.MSELoss()
optimizer = torch.optim.Adam(snn.parameters(), lr=0.0033)

def train_for_epoch(model, optimizer, loader):
    correct = 0
    total = 0
    running_loss = 0
    start_time = time.time()
    criterion = nn.MSELoss()
    for i, (images, labels) in enumerate(loader):
        model.zero_grad()
        optimizer.zero_grad()

        images = images.float().to(device)
        outputs = snn(images)
        labels_ = torch.zeros(batch_size, 10).scatter_(1, labels.view(-1, 1), 1)
        loss = criterion(outputs.cpu(), labels_)
        _, predicted = outputs.cpu().max(1)
        total += float(labels.size(0))
        correct += float(predicted.eq(labels).sum().item())
        running_loss += loss.item()
        loss.backward()
        optimizer.step()

    return running_loss / len(loader), 100 * correct / total

def eval_on_validation(model, loader):
    running_loss = 0
    correct = 0
    total = 0
    criterion = nn.MSELoss()

    with torch.no_grad():
        for batch_idx, (inputs, targets) in enumerate(loader):
            inputs = inputs.to(device)
            optimizer.zero_grad()
            outputs = model(inputs)
            labels_ = torch.zeros(batch_size, 10).scatter_(1, targets.view(-1, 1), 1)
            loss = criterion(outputs.cpu(), labels_)
            _, predicted = outputs.cpu().max(1)
            total += float(targets.size(0))
            correct += float(predicted.eq(targets).sum().item())
            running_loss += loss.item()
            if batch_idx % 100 == 0:
                acc = 100. * float(correct) / float(total)
                print(batch_idx, len(loader), ' Acc: %.5f' % acc)

    return running_loss / len(loader), 100 * correct / total

n_epochs = 1
optimizer = lr_scheduler(optimizer, n_epochs, 0.0033, 40)

for i in range(n_epochs):
    torch.cuda.synchronize(device=device)
    since = int(round(time.time() * 1000))
    torch.manual_seed(3)
    torch.cuda.manual_seed_all(44)

    train_loss, train_accuracy = train_for_epoch(snn, optimizer, train_loader)
    train_losses.append(train_loss)
    train_accs.append(train_accuracy)
    torch.cuda.synchronize(device=device)
    time_elapsed = int(round(time.time() * 1000)) - since
    print('training time elapsed {}ms'.format(time_elapsed))

    # Calculating time for testing
    torch.cuda.synchronize(device=device)
    tsince = int(round(time.time() * 1000))
    val_loss, val_acc = eval_on_validation(snn, val_loader)
    torch.cuda.synchronize(device=device)
    ttime_elapsed = int(round(time.time() * 1000)) - tsince

    print('test time elapsed {}ms'.format(ttime_elapsed))
    is_best = val_acc > best_acc
    best_acc = max(val_acc, best_acc)
    PATH = '/Users/Tehreem/Desktop/saved_mnist_model/state_dict_model.pt'
    torch.save(snn.state_dict(), PATH)

    val_losses.append(val_loss)
    val_accs.append(val_acc)
    print(i + 1, " / ", n_epochs, ": loss: ", train_losses[-1], "Accuracy:", train_accs[-1], ", val_loss: ",
          val_losses[-1], ", val_accs: ",
          val_accs[-1])

dataiter = iter(val_loader)
data, labels = dataiter.next()
data, labels = data.to(device), labels.to(device)

#print(data.shape)
y = labels.data[0]
plt.figure(figsize=(12, 12))

img = data.cpu().numpy()[0, :, :, :].squeeze()
plt.imshow(img)
plt.title("Input image x of label y={}:".format(y))
plt.show()

# plotting neuron's activations:
snn.visualize_all_neurons(img)
snn.visualize_neuron(img, layer_idx=0, neuron_idx=0)

**Error**

![image](https://user-images.githubusercontent.com/38427571/94006373-b3d4ff80-fdda-11ea-97b8-d12813a02070.png)

Thanks
